### PR TITLE
merge pr-downloader branch

### PR DESCRIPTION
**Pull request by abma**
thoughts about this?

this would make pr-downloader mandatory, atm its optional. is this ok or should it still be optional?

the code still needs some improvements, but the api / file structure should stay the same.

### CODE REVIEW

**Comment by gajop** [comment]
HashLobbyPassword doesn't really belong in LuaVFSDownloand and can be merged soon independently.

**Comment by abma** [comment]
agreed, i've removed it.

File: rts/Lua/LuaHandle.cpp [hunk 1]
```diff
@@ -1614,6 +1615,8 @@ void CLuaHandle::Update()
 
 	// call the routine
 	RunCallIn(L, cmdStr, 0, 0);
+
+	LuaVFSDownload::ProcessDownloads();
```
Comment (ashdnazg) [inline rts/Lua/LuaHandle.cpp]: Is this supposed to happen after every Update of every LuaHandle or just once every Update event?

Comment (abma) [inline rts/Lua/LuaHandle.cpp]: its meant to be called just once every Update event. not sure how often/when LuaHandle::Update is called.

Comment (ashdnazg) [inline rts/Lua/LuaHandle.cpp]: you can make LuaVFSDownload an eventClient and then change ProcessDownloads() to Update()

Comment (abma) [inline rts/Lua/LuaHandle.cpp]: done!

Comment (ashdnazg) [inline rts/Lua/LuaHandle.cpp]: superb!

**Comment by rtri** [comment]
Erhm, why is a branch with this many TODO's and FIXME's being considered for merging?

**Comment by abma** [comment]
- its basicly "modular", all relevant code is in one file, VFSDownload.cpp (+pr-downloader submodule)
- it works with some restrictions (i.e. when download is running, no new dl can be started)
- fixing all TODO's / FIXME's will take a lot of time and this with the risk that its not useful / won't be merged at all

**Comment by rtri** [comment]
What kind of help exactly? What e.g. would be the specific difficulties with cleaning up at least the FIXME's?

This work dates back a year (looking at the commits), so why can more time not be taken to include clean non-WIP code?

**Comment by abma** [comment]
> What kind of help exactly?

feedback like ashdnazg did or maybe direct help/hints for the FIXME/TODOs.

> What e.g. would be the specific difficulties with solving at least the FIXME's?
lack of c++ knowledge, lack of knowledge of spring internals and lack of time.

> This work dates back a year (looking at the commits), so why can more time not be taken to include clean non-WIP code?

a to long time-scale for such a project mostly will make it die when no "release" is made. this branch contains changes which possible helps with an ingame-lobby / missions.

**Comment by rtri** [comment]
"I lack knowledge and time, please help me or I merge" is project blackmail, sorry. I know what the branch contains and do understand the desire/motivation to finish it, but owner-PR's should not be used as a pressure tool to put work on others who might not have time or knowledge either.

**Comment by abma** [comment]
> "I lack knowledge and time, please help me or I merge" is project blackmail

this pull request is meant for code review + feedback.

>  but owner-PR's should not be used as a pressure tool to put demands on others who might not have time or knowledge either.

what is owner-pull-requests? if they have no time nor knowledge, they should ignore it (like the other pull requests)

using pull requests for review / discuss isn't this good, but its the best thing i currently know about what can be used for this.

**Comment by ashdnazg** [comment]
Personally I see this PR as a convenient way to watch/review the branch development and discuss it.
Never thought of it as "please help me or I merge".

**Comment by gajop** [comment]
I think there's no rush in merging this, but I would prefer to have this PR open for communication. Personally it's not a priority for me.

**Comment by gajop** [comment]
Would like an updated review on this. I think the only two thing that are left to be implemented (for the most basic functionality) should be done on the pr-downloader side.

**Comment by abma** [comment]
sorry for the delay. looks good, still needs some work but should be usable and not conflict with the rest of the engine when not used -> will merge soon when no new big issues are found / pointed out

**Comment by gajop** [comment]
Sounds great. There's no rush, but it would be great if it gets in for 101.
It'll make lobby development easier and I don't think it's likely to cause
any engine issues.

File: rts/System/SpringApp.cpp [hunk 2]
```diff
@@ -901,6 +903,7 @@ void SpringApp::Reload(const std::string& script)
 	// PreGame allocates clientNet, so we need to delete our old connection
 	SafeDelete(clientNet);
 
+	SafeDelete(luavfsdownload);
```
Comment (ashdnazg) [inline rts/System/SpringApp.cpp]: needs to be deleted in SpringApp::ShutDown too

**Pull request merged**